graphql/src/query/mod.rs

```diff
@@ -66,17 +64,33 @@ where
         deadline: options.deadline,
     };
 
-    let result = match *operation {
-        // Execute top-level `query { ... }` expressions
-        q::OperationDefinition::Query(q::Query {
-            ref selection_set, ..
-        }) => execute_root_selection_set(&ctx, selection_set, &None),
+    let result = match operation {
+        // Execute top-level `query { ... }` and `{ ... }` expressions.
+        q::OperationDefinition::Query(q::Query { selection_set, .. })
+        | q::OperationDefinition::SelectionSet(selection_set) => {
+            let complexity = ctx.root_query_complexity(
+                sast::get_root_query_type_def(&ctx.schema.document).unwrap(),
+                selection_set,
+            );
 
-        // Execute top-level `{ ... }` expressions
-        q::OperationDefinition::SelectionSet(ref selection_set) => {
-            execute_root_selection_set(&ctx, selection_set, &None)
-        }
+            info!(
+                query_logger,
+                "Execute query";
+                "query" => query.document.format(&Style::default().indent(0)).replace('\n', " "),
+                "complexity" => format!("{:?}", complexity),
+            );
 
+            match (complexity, options.max_complexity) {
+                (Err(e), _) => Err(vec![e]),
+                (Ok(complexity), Some(max_complexity)) if complexity > max_complexity => {
+                    Err(vec![QueryExecutionError::TooComplex(
```

Comment:
I'm realizing it would be worth logging query errors further down where we log `Finished query`. That way we capture them in the node logs and not just on the client (which often, we don't have access to).